### DISCUSSION TITLE
Unable to find context directory when deploying via compose

### DISCUSSION THREAD

**Discussion by paulrobinsontkd** [louislam/dockge#277] [Ask for Help]
### ⚠️ Please verify that this bug has NOT been reported before.

- [X] I checked and didn't find similar issue

### 🛡️ Security Policy

- [X] I agree to have read this project [Security Policy](https://github.com/louislam/dockge/security/policy)

### Description

I am attempting to deploy a docker container from https://github.com/eko/pihole-exporter

```
[+] Running 1/1
 ✔ Container pihole-exporter  Started                                                               0.0s 
[+] Running 1/1
 ! pihole-exporter Warning                                                                          1.3s 
[+] Building 0.0s (0/0)                                                                   docker:default
unable to prepare context: path "/share/docker/build/pihole-exporter/" not found
```

Docker compose file

```
version: "3.6"
services:
  pihole-exporter:
    build:
      context: /share/docker/build/pihole-exporter/
      dockerfile: /share/docker/build/pihole-exporter/Dockerfile
    image: ekofr/pihole-exporter:amd64
    container_name: pihole-exporter
    ports:
      - 9617:9617/tcp
    expose:
      - 9617
    environment:
      PIHOLE_HOSTNAME: 192.168.1.200,192.168.1.201
      PIHOLE_PORT: 80
      PIHOLE_PASSWORD: XXXXX
      INTERVAL: 30s
      PORT: 9617
    networks:
      qnet-static-eth4:
        ipv4_address: 192.168.10.202
    restart: always
networks:
  qnet-static-eth4:
    external: true
```

If i try to deploy via dockge it is unable to find the directories for context and dockerfile

However if i deploy via the OS command line i have no issues

```
[admin@qnap pihole-exporter]# docker compose -f compose.yaml up -d
[+] Running 0/1
 ⠿ pihole-exporter Warning                                                                                         1.7s
[+] Building 0.9s (14/14) FINISHED
 => [internal] load build definition from Dockerfile                                                               0.0s
 => => transferring dockerfile: 84B                                                                                0.0s
 => [internal] load .dockerignore                                                                                  0.0s
 => => transferring context: 2B                                                                                    0.0s
 => [internal] load metadata for docker.io/library/golang:1.21.5-alpine3.17                                        0.8s
 => [auth] library/golang:pull token for registry-1.docker.io                                                      0.0s
 => [stage-1 1/2] WORKDIR /root/                                                                                   0.0s
 => [internal] load build context                                                                                  0.0s
 => => transferring context: 3.55kB                                                                                0.0s
 => [builder 1/6] FROM docker.io/library/golang:1.21.5-alpine3.17@sha256:92cb87af996ec6befc85f0aec27e12ead2fab396  0.0s
 => CACHED [builder 2/6] WORKDIR /go/src/github.com/eko/pihole-exporter                                            0.0s
 => CACHED [builder 3/6] COPY . .                                                                                  0.0s
 => CACHED [builder 4/6] RUN apk --no-cache add git alpine-sdk                                                     0.0s
 => CACHED [builder 5/6] RUN GO111MODULE=on go mod vendor                                                          0.0s
 => CACHED [builder 6/6] RUN CGO_ENABLED=0 GOOS=$OS GOARCH=$ARCH go build -ldflags '-s -w' -o binary ./            0.0s
 => CACHED [stage-1 2/2] COPY --from=builder /go/src/github.com/eko/pihole-exporter/binary pihole-exporter         0.0s
 => exporting to image                                                                                             0.0s
 => => exporting layers                                                                                            0.0s
 => => writing image sha256:d3c26371bde522987fba618c4aab12dd7692c5facf9a90bd3ddbcb57b132ecbc                       0.0s
 => => naming to docker.io/ekofr/pihole-exporter:amd64                                                             0.0s
[+] Running 1/1
 ⠿ Container pihole-exporter  Started                                                                              0.5s
[admin@qnap pihole-exporter]#
```

### 👟 Reproduction steps

Deployed dcoker compose file with build and context parameters

### 👀 Expected behavior

Deploy contain and build custom image

### 😓 Actual Behavior

Fails to find context directory and will not bring up container

### Dockge Version

1.2.0

### 💻 Operating System and Arch

Linux qnap 5.10.60-qnap #1 SMP Tue Sep 26 01:46:28 CST 2023 x86_64 GNU/Linux

### 🌐 Browser

N/A

### 🐋 Docker Version

Docker version 20.10.22-qnap7, build 57ed8b8

### 🟩 NodeJS Version

Unknown

### 📝 Relevant log output

```shell
[+] Running 1/1
 ✔ Container pihole-exporter  Started                                                               0.0s 
[+] Running 1/1
 ! pihole-exporter Warning                                                                          1.3s 
[+] Building 0.0s (0/0)                                                                   docker:default
unable to prepare context: path "/share/docker/build/pihole-exporter/" not found
```

**Comment by louislam** [comment]
Dockge is a container too, anything outside the `stacks` folder or `data` folder are not accessible.

**Reply by paulrobinsontkd** [reply]
Ahhh yes i see.

Would it work if i add my build (context and dockerfile) directory as a volume to dockge?

**Reply by louislam** [reply]
I think yes if the build script only used this build folder.

**Reply by paulrobinsontkd** [reply]
Yes i can confirm it works by adding this as a volume to dockge.

I have a few more custom containers to try so hopefully this will work for all the others as well.

Thanks for replying and confirming the underlying issue.  Great work and thanks for creating this application

**Reply by oopsunix** [reply]
I'm trying to use dockge to deploy the h5ai container from [awesometic/h5ai] (https://github.com/awesometic/docker-h5ai), where it needs to volumes follows:

```
volumes:
      - /root/share:/h5ai
      - ./config:/config
```

When I used dockge for deployment, he didn't create the config folder, and even if I created the config folder manually, it couldn't map the files.

But when I execute the'docker compose up -d 'command in ssh, he can successfully create and map the file. Why is that?

H5ai composes files in：/root/docker/h5ai/compose.yml

Dockge compose:

```
version: "3.8"
services:
  dockge:
    image: louislam/dockge:1
    container_name: dockge
    restart: unless-stopped
    ports:
      # Host Port : Container Port
      - 45001:5001
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - ./data:/app/data
      - /root/docker:/opt/stacks
    environment:
      # Tell Dockge where is your stacks directory
      - DOCKGE_STACKS_DIR=/opt/stacks
```

**Comment by trythatagain** [comment]
To me that looks like a permissions thing. What user do ssh in with? If it's root then that might be the problem.

**Comment by JinbaIttai** [comment]
as @trythatagain says it could be a permissions error - make sure the userid you configured in h5ai (with the UID environment variable) has permission to write to /root/docker/h5ai/config

But also, you're passing through a relative path to dockge with your ` ./config:/config` which dockge is translating to `$DOCKGE_STACKS_DIR/h5ai/config` 

ie `/opt/stacks/h5ai/config`

But on the host the real directory is `/root/docker/h5ai/config` so docker can't find it.

This is why it's very important to make sure your stacks directory is the same on the host and inside dockge

and is why there is a big warning in the sample [compose.yaml](https://github.com/louislam/dockge/blob/master/compose.yaml)

You really should change your Dockge compose file to have /root/docker:/root/docker and DOCKGE_STACKS_DIR=/root/docker
